### ISSUE TITLE
Auto-delete branch on pull request merge

### ISSUE THREAD

**Issue by stuartpb** [isaacs/github#1121]
It would be nice if we could have a checkbox / button or something for pull requests to make it so, when the pull request is merged, rather than having to revisit the PR and manually click the "Delete branch" button, the branch for the pull request would be deleted automatically. This is one of the features @jessfraz was just tweeting about running out of GitHub API calls to implement for herself: https://twitter.com/jessfraz/status/930154061869002752

Just sent this suggestion to GitHub Support.

**Comment by stuartpb** [comment]
Wow, that was fast, looks like they've got a template response email for feature requests:

> Hi there,
>
> Thanks for reaching out! I'd be happy to pass this along to the team to consider as a feature request. I can't promise if or when it will be implemented, but your suggestion is definitely in the right hands!
>
> Regards,
> Laura
> @lecoursen
> GitHub Support

**Comment by pedro93** [comment]
Amazing!

**Comment by davyengone** [comment]
Is this feature request still pending? That would be great to have it

**Comment by martinbonnin** [comment]
For what it's worth, I worked around this issue by writing a small script that auto-deletes merged branches every night. Not ideal but works.

**Comment by jimbol** [comment]
Would you be willing to share that with us @martinbonnin? I'd `star` that repo

**Comment by ericanderson** [comment]
This just annoyed me recently enough that I wrote a script. I intend to produce other useful things for working against git/gh inside it and PRs are welcome. Just ran it on internal instance to success: https://github.com/ericanderson/ghouls

**Comment by treps** [comment]
Have there been any updates on whether this is going to be a thing anytime soon?

**Comment by tonypiazza** [comment]
@lecoursen This would be very helpful. Can you give us an update on this?

**Comment by monperrus** [comment]
I guess that this can be implemented as a webhook. Has anybody ever done so?

**Comment by sparky005** [comment]
+1 for this feature.

**Comment by tansaku** [comment]
we'd love it too!

**Comment by renoirfaria** [comment]
Hey guys, what about the following announcements and help articles? It seems that this feature already existed and someday and simply being vanished?

- https://blog.github.com/2012-12-03-tidying-up-after-pull-requests/
- https://help.github.com/articles/deleting-and-restoring-branches-in-a-pull-request/

Confused.

**Comment by wareczek** [comment]
@renoirfaria read carefully ;) This issue is about automatically delete branch after merge. Not click on button action.

**Comment by AraHaan** [comment]
Not to mention it would be nice to be able to physically delete issues and pull requests. I know it is a simple curl delete request away but I do not get the actual params I need to delete invalid issues/pull requests manually nor do I always have the ability as I have to use my iPhone most of the time on github unfortunately.

**Comment by renoirfaria** [comment]
@wareczek right, but both alternatives (automatic and button deletion) seems to be are no longer available.

**Comment by wareczek** [comment]
No. Button is still there after merge. You can delete or restore branch.

**Comment by thomas-scrace** [comment]
+1

**Comment by cgirard-doc** [comment]
+1

**Comment by AraHaan** [comment]
I just got the idea. One could make a CI bot for their org that can detect pr merge and then check if the base branch is in the same repository, if so then delete it but then the CI would have to have means of editing the repository itself for that though. But at least it would be something. However like github pages it would be nice if only could setup a folder or whatever under branch master that then would run the ci itself automatically using Github as the host. For those who cannot run it on their devices and do not have money for a VPS or something to run it elsewhere. No programming language preference on this comment though so any language that can be used to make a Github CI is ok.

**Comment by SvanBoxel** [comment]
You can install this GitHub app that removes a branch after the PR gets merged. https://github.com/apps/delete-merged-branch

**Comment by tiberiomoliveira** [comment]
Maybe this will do the job: https://developer.github.com/v3/git/refs/#delete-a-reference

**Comment by SvanBoxel** [comment]
Or this https://probot.github.io/apps/delete-merged-branch/

**Comment by Jumhyn** [comment]
+1

**Comment by Oyelaking** [comment]
+1

**Comment by 27cm** [comment]
+1

**Comment by AraHaan** [comment]
@SvanBoxel it works, but sadly that seems more of a hack than anything. It would be nice if that bot was automatically installed on any *newly* made repository from now on and does not require a per-user installation unless it is for enabling it on a older or old repository.

**Comment by jd** [comment]
You could also leverage a [Mergify](https://mergify.io) rule, see:
https://medium.com/mergify/keep-your-repository-clean-delete-your-merged-branches-4b779f4848ae

**Comment by mspiegel** [comment]
I've implemented this feature as a GitHub App. Looking for beta-testers. Check out https://github.com/apps/always-be-closing if you're interested. Any feedback would be appreciated at https://github.com/thoughtdealership/abc-users.

**Comment by amueller** [comment]
there is a button right after the merge button. That seems like enough, right?

**Comment by clarkbw** [comment]
Auto-delete branches 🎉 

Check it out! https://github.blog/changelog/2019-07-31-automatically-delete-head-branches-of-pull-requests/

**Comment by mgoddard27** [comment]
Will the 'auto-delete branches' feature be available on GitHub Enterprise as well, anytime soon?

**Comment by SvanBoxel** [comment]
@mgoddard27 Assuming you mean GitHub Enterprise Server; yes. Normally we (GitHub) ship features to Enterprise Server after we battle-tested them on GitHub.com. This could take a couple of months because of the quarterly release cadence.

**Comment by testworksau** [comment]
@clarkbw any idea when an API to turn this setting on / off might arrive? It's not listed here:
https://developer.github.com/v3/repos/#edit

**Comment by clarkbw** [comment]
I don't have a timeline for the API yet, looking into it.

> On another note - a repository maintainer has access to enable / disable the merge button to allow merge commits / squash merging / rebase merging in the GitHub UI, but does not have the ability to see the new `Automatically delete head branches` setting. Was this intentional?

re: the Maintainer role?  I'll look into it.

**Comment by testworksau** [comment]
@clarkbw I've found that this setting is unexpectedly being turned off on us when using the Edit repos endpoint.

We use the https://developer.github.com/v3/repos/#edit API to configure merge strategies on our repositories. We typically send the following information (and only the following information) in the patch request:

```
{
  "allow_merge_commit": false,
  "allow_squash_merge": true,
  "allow_rebase_merge": true
}
```

After calling the API, the `Automatically delete head branches` setting turns off.

**Comment by clarkbw** [comment]
That’s not good...  can you send this through support so it gets to the right people?  Thx!

**Comment by meabed** [comment]
> @clarkbw I've found that this setting is unexpectedly being turned off on us when using the Edit repos endpoint.
> 
> We use the https://developer.github.com/v3/repos/#edit API to configure merge strategies on our repositories. We typically send the following information (and only the following information) in the patch request:
> 
> ```
> {
>   "allow_merge_commit": false,
>   "allow_squash_merge": true,
>   "allow_rebase_merge": true
> }
> ```
> 
> After calling the API, the `Automatically delete head branches` setting turns off.

I have exactly the same issue, @clarkbw there is no option to toggle this through the API as well.
![image](https://user-images.githubusercontent.com/45731/68537353-7d85c180-037b-11ea-86e3-8aeba5d54696.png)

![image](https://user-images.githubusercontent.com/45731/68537378-f84edc80-037b-11ea-8b0b-374762482be3.png)

It would be great if this could be addressed :)

**Comment by terowz** [comment]
do you have any idea when this will be available in the API as well?
or am I just missing the docs on it?

**Comment by testworksau** [comment]
The bug has finally been fixed, and this parameter is now available through the API, as `delete_branch_on_merge`.

https://developer.github.com/v3/repos/#edit

As a bonus for those using PyGitHub, I've also raised a PR to enable this parameter to be configured, which has been merged and is available in PyGitHub 1.46

https://github.com/PyGithub/PyGithub/releases/tag/v1.46
